process/process_linux.go

```diff
@@ -49,6 +90,112 @@ type MemoryMapsStat struct {
 	Swap         uint64 `json:"swap"`
 }
 
+func (m *MemoryMapsStat) updateAttribute(fields []string) error {
+	if len(fields) < 2 {
+		return fmt.Errorf("update attribute not enough fields")
+	}
+	var v *uint64
+	switch fields[0] {
+	case "Size:":
+		v = &m.Size
+	case "Rss:":
+		v = &m.Rss
+	case "Pss:":
+		v = &m.Pss
+	case "Shared_Clean:":
+		v = &m.SharedClean
+	case "Shared_Dirty:":
+		v = &m.SharedDirty
+	case "Private_Clean:":
+		v = &m.PrivateClean
+	case "Private_Dirty:":
+		v = &m.PrivateDirty
+	case "Referenced:":
+		v = &m.Referenced
+	case "Anonymous:":
+		v = &m.Anonymous
+	case "Swap:":
+		v = &m.Swap
+	default:
+		// ignore other fields
+		return nil
+	}
+
+	t, err := strconv.ParseUint(fields[1], 10, 64)
+	if err != nil {
+		return err
+	}
+
+	*v = t
+	return nil
+}
+
+/*
+   Parse a memory section from /proc/pid/maps or header section line of /proc/pid/smaps
+
+ StartAddr        EndAddr    Permission Offset Device Inode              Path
+ffffffffff600000-ffffffffff601000 --xp 00000000 00:00 0                  [vsyscall]
+*/
+func (ms *MemoryMapsStat) parseHeader(fields []string) error {
+	if len(fields) < 5 {
+		return fmt.Errorf("can't parse section, not enough fields '%v'", fields)
+	}
+
+	/* Address Range */
+	i := strings.IndexRune(fields[0], '-')
+	if i < 0 {
+		return fmt.Errorf("can't parse address section, not enough fields '%v'", fields[0])
+	}
+	startAddr := fields[0][:i]
+	endAddr := fields[0][i+1:]
+	a, err := strconv.ParseUint(startAddr, 16, 64)
+	if err != nil {
+		return err
+	}
+	ms.StartAddr = uintptr(a)
+	a, err = strconv.ParseUint(endAddr, 16, 64)
+	if err != nil {
+		return err
+	}
+	ms.EndAddr = uintptr(a)
+
+	// Permission
+	if err := ms.Permission.parse(fields[1]); err != nil {
+		return err
+	}
+
+	// Offset
+	a, err = strconv.ParseUint(fields[2], 16, 64)
+	if err != nil {
+		return err
+	}
+	ms.Offset = a
+
+	// Device
+	dev := fields[3]
+	if len(dev) < 5 || !strings.Contains(dev, ":") {
+		return fmt.Errorf("can't parse device section, '%s'", dev)
+	}
+	ms.Device = dev
+
+	// Inode
+	a, err = strconv.ParseUint(fields[4], 10, 64)
+	if err != nil {
+		return err
+	}
+	ms.Inode = a
+
+	if len(fields) >= 5 {
```

Comment:
At this point, you know that `fields` is at least of length 5. Can you remove this if?

Reply:
we are not sure that we have at least 5 fields here, as anonymous map have no such field set

 StartAddr        EndAddr    Permission Offset Device Inode              Path
ffffffffff600000-ffffffffff601000 --xp 00000000 00:00 0                  [vsyscall]

On anonymous map Path is empty

Reply:
I fixup the "if len(fields) >= 5 {" test too.   Thanks